spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,335 @@
+* To improve CLI ergonomics, could implement additive `--allow-resources`/`--allow-modules` args (current flags replace full list)
+* Manage external reader processes separately from EvaluatorImpl lifetime
+** Potential large savings in per-evaluator overhead for Java API and Language Binding usage
+** Savings for CLI usage (primary use case) would be minimal
+** Code is more complicated (need an ExternalReaderManager sort of mechanism tracking unique commands => processes)
+** Change could be made as followup work with only changes to Java APIs and internals
+
+== Alternatives considered
+
+=== One shot, per-read subprocesses
+
+Instead of "persistent" reader processes invoked during evaluator initialization.
+Instead of using the msgpack message-passing API, reader binaries could be invoked with the read URI as a CLI argument and return their result on standard output.
+This potentially greatly lowers the barrier to entry for implementing external readers, even allowing them to be implemented by shell scripts.
+
+This approach does not have a clean way to support globbed reads.
+To resolve globs, Pkl can require many list modules/resources requests.
+It's not clear one-shot reader processes would be invoked differently to distinguish read requests from list requests.
+Multiple invocations would also have potentially significant overhead, especially for readers implemented in interpreted languages.
+
+There is definitely value in supporting significantly reduced barrier to reader implementation, especially when globbing is not required.
+One way this gap might be closed is with a "shim" reader process that translates the message passing API calls to subprocess invocations:
+
+[source,text]
+----
+$ pkl eval <module> --external-reader 'pkl-cmd ldap=pkl-ldap.sh' --allowed-resources prop:,ldap:
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+It may even make sense for the `pkl` binary itself to provide this functionality.
+
+=== Up-front scheme -> reader registration
+
+Instead of Pkl starting reader subprocesses and discovering supported schemes during evaluator initialization, an alternative approach would be to explicit register this mapping.
+This would allow reader processes to be launched on first read instead of during evaluator initialization.
+This is more efficient in cases where the reader is not actually needed, but requires a greater amount of up-front configuration, especially when the same reader executable will be used for multiple schemes.
+
+[source,text]
+----
+$ pkl eval <module> --external-resource-reader ldap=pkl-ldap --allowed-resources prop:,ldap:
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+This approach raises a few questions:
+
+* Should declaring an external reader automatically allow reads for its scheme?
+Declaring explicit allowed resources should probably disable this behavior.
+* What happens when a reader doesn't support the scheme it is declared for?
```

Comment:
What does this mean? Is this when the reader sends back an error message in its `Read {Module,Resource} Response`? If so, that should be bubbled back up as a Pkl Error during evaluation.

Reply:
Taking the `pkl-ldap` reader described here as an example: it answers for resources with the `ldap` and `ldaps` schemes. If a user makes a mistake and configures the read eg. as `--external-resource-reader postgresql=pkl-ldap` then the reader process can receive messages with a scheme it doesn't support. The discovery-driven design can return the native Pkl "no reader registered for scheme error", while explicit registration means the request is still sent to the external reader where the returned error is dependent on the client-side implementation which hopefully can handle this gracefully and return a helpful error.